src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -371,43 +426,126 @@ public int length(DynamicObject matchData) {
     public abstract static class PreMatchNode extends CoreMethodArrayArgumentsNode {
 
         @Child private TaintResultNode taintResultNode = new TaintResultNode();
+        @Child private RopeNodes.MakeSubstringNode makeSubstringNode = RopeNodes.MakeSubstringNode.create();
 
-        @Specialization
+        public static PreMatchNode create() {
+            return PreMatchNodeFactory.create(new RubyNode[0]);
+        }
+
+        public abstract DynamicObject execute(DynamicObject matchData);
+
+        @Specialization(guards = "isConstructed(matchData)")
+        public Object preMatchFast(DynamicObject matchData) {
+            return Layouts.MATCH_DATA.getPre(matchData);
+        }
+
+        @Specialization(guards = "!isConstructed(matchData)")
         public Object preMatch(DynamicObject matchData) {
-            return taintResultNode.maybeTaint(Layouts.MATCH_DATA.getSource(matchData), Layouts.MATCH_DATA.getPre(matchData));
+            DynamicObject source = Layouts.MATCH_DATA.getSource(matchData);
+            Rope sourceRope = StringOperations.rope(source);
+            Region region = Layouts.MATCH_DATA.getRegion(matchData);
+            int start = 0;
+            int length = region.beg[0];
+            Rope rope = makeSubstringNode.executeMake(sourceRope, start, length);
+            DynamicObject newStr = createSubstring(source, rope);
+            Layouts.MATCH_DATA.setPre(matchData, newStr);
```

Comment:
It seems MRI always return different String on `$1`, `$'`, `$~.post_match`, etc.
So I think we should cache the Rope if we want, but not the String (it would bad if the String is modified).
And then it might actually be useless to cache the Rope since we can create a SubstringRope very efficiently so then we can just stop caching and have a single fast path :)

But this is something which was already wrong before, so I think it's better to do it in another PR.

Reply:
Wow, some testing suggests MRI just keeps the info on a match to recreate all the information every time. We could easily ditch pre, post, match, and values from the match data layout.

Reply:
Exactly, I think that's best :)
The only concern I have is it might be expensive to recompute a SubstringRope for non-7-bit Strings, since then a CodeRange is needed (to know if the substring is 7-bit or just VALID if the whole string is VALID).